File: Algorithms/Math/python/count_primes.py
class Solution:
    def countPrimes(self, n: int) -> int:
        """
        Args:
            n: a non-negative number.

        Returns:
            the number of prime less than n.

        Example:
            Input: 10
            Output: 4

        Idea of solution:
            Use Sieve of Eratosthenes algorithm.
        
        Time complexity: O(n)
        Space complexity: O(n)
        """
        if n < 2: return 0
        is_prime = [False if i < 2 else True for i in range(n)]
        for i in range(2, int(n ** 0.5) + 1):
            if is_prime[i]:
                for j in range(i * i, n, i):
                    is_prime[j] = False
        return is_prime.count(True)

File: Algorithms/Math/java/CountPrimes.java
/**
 * Count the number of prime numbers less than a non-negative number, n.
 * 
 * Idea of solution: Use Sieve of Eratosthenes alogorithm
 * 
 * Time Complexity: O(n), Space Complexity: O(n)
 */
public class CountPrimes {
    /**
     * @param n: a non-negative number.
     * @return the number of prime less than n.
     */
    public int countPrimes(int n) {
        // Initialize
        boolean[] isPrime = new boolean[n];
        for (int i = 2; i < isPrime.length; i++) {
            isPrime[i] = true;
        }
        // Loop over to update
        for (int i = 2; i < (int) Math.sqrt(n) + 1; i++) {
            if (isPrime[i]) {
                for (int j = i * i; j < n; j += i) {
                    isPrime[j] = false;
                }
            }
        }
        // Count number of prime number
        int res = 0;
        for (boolean p : isPrime) {
            if (p)
                res++;
        }
        return res;
    }
}